### PR TITLE
refactor: use htmlentities to avoid issues with html elements in metadata

### PR DESCRIPTION
**Pull request by alfonsobries**
## Summary

Dependant of: https://github.com/ArkEcosystem/marketsquare.io/pull/2220

https://app.clickup.com/t/wkx09q

- Adds a new component that will help us to add meta-titles that are also trimmed (and xss safe)
- Replaces the need to use trick like this `<title>{{ trim(View::yieldContent('title', trans("metatags.home.title"))) }}</title>` to trim the titles that is the problem behind the issue on the ticket (see msq PR).

#### Before:
image

#### After:
image

## Checklist

-   [x ] I checked my UI changes against the design and there are no notable differences
-   [x] I checked my UI changes for any responsiveness issues
-   [x] I checked my (code) changes for obvious issues, debug statements and commented code
-   [x] I provided a screenshot of my changes to the component _(if applicable)_
-   [ ] I regenerated the `icons.html` file and checked if my newly added icon is shown correctly _(if necessary)_
-   [ ] I added an explanation on how to use the component to the readme _(if necessary)_
-   [ ] Documentation _(if necessary)_
-   [ ] Tests _(if necessary)_
-   [ ] Ready to be merged

### CODE REVIEW

File: resources/views/metadata-title.blade.php [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{{-- Uses a regular `echo` expression since blade syntax creates unexpected HTML
+entities like `&#039;` that doesnt work correctly when shared for example. --}}
+<title><?php echo htmlentities(trim($slot)) ?></title>
```
Comment (ItsANameToo) [inline resources/views/metadata-title.blade.php:3]: Can you explain why the `title` tag is set with this? the issue appears in the meta title, not the regular title tag 🤔 
image

File: src/UserInterfaceServiceProvider.php [hunk 2]
```diff
@@ -247,6 +247,7 @@ private function registerBladeComponents(): void
         Blade::component('ark::loader-icon', 'ark-loader-icon');
         Blade::component('ark::message', 'ark-message');
         Blade::component('ark::metadata', 'ark-metadata');
+        Blade::component('ark::metadata-title', 'ark-metadata-title');
```
Comment (ItsANameToo) [inline src/UserInterfaceServiceProvider.php:250]: there is already a metadata-tags component that sets metadata data, why add a new component instead of adjusting the existing one? 🤔

**Review by ItsANameToo — CHANGES_REQUESTED** [review]

File: resources/views/metadata-title.blade.php (hunk 1 above)
Comment (alfonsobries) [reply resources/views/metadata-title.blade.php:3]: @ItsANameToo mm... the problem is (potentially) in all the meta (including regular `title`) (Ill use the same solution to the rest of the metatags)

The reason: we add the meta data like this (`View::yieldContent('meta-title', trans('metatags.home.title'))`) and the we use the blade `{{ }}` directive to then write the content `<meta name="description" content="{{ trim($description) }}" />` and the problem is that the blade directive ( `{{ }}` ) encodes the chars.

That's an alternative to the regular `@yield` directive and I didn't add those components but I am pretty sure the reason was to trim the texts and extract the components.

File: src/UserInterfaceServiceProvider.php (hunk 2 above)
Comment (alfonsobries) [reply src/UserInterfaceServiceProvider.php:250]: @ItsANameToo yeah I can reuse the same component, will fix in a moment

**Comment by alfonsobries** [comment]
@ItsANameToo ignore my previous comments the solution was working but the main problem still was on msq.

I updated this PR to adds the `title` to the metatags component (a component is still needed to trim the msq titles), the rest of the fixes are no on the msq PR

File: resources/views/metadata-tags.blade.php [hunk 3]
```diff
@@ -1,12 +1,15 @@
 @props([
     'title',
     'description',
+    'metaTitle' => null,
     'image' => null,
 ])
 
+<title>{{ $title }}</title>
```
Comment (ItsANameToo) [inline resources/views/metadata-tags.blade.php:8]: why was this needed? Title has no issues for me on that blog 🤔 
image

**Review by ItsANameToo — CHANGES_REQUESTED** [review]

**Comment by ItsANameToo** [comment]
also think your before/after images are switched @alfonsobries

File: resources/views/metadata-tags.blade.php (hunk 3 above)
Comment (alfonsobries) [reply resources/views/metadata-tags.blade.php:8]: @ItsANameToo is that local?, the production site has the issue
image

`view-source:https://marketsquare.io/projects/marketsquare/blog/i-spent-6-months-inside-the-metaverse-and-heres-why-it-needs-crypto`

Comment (alfonsobries) [reply resources/views/metadata-tags.blade.php:8]: and I have it locally (before the patch)

Comment (ItsANameToo) [reply resources/views/metadata-tags.blade.php:8]: it's on production, but in devtools. looks like `view-source:` encodes it :sus:

Comment (alfonsobries) [reply resources/views/metadata-tags.blade.php:8]: @ItsANameToo interesting... seems that devtools encodes text inside nodes but not regular attributes like the `og:title` I suppose is because is what you usually see

Ill check if the title alone affect the sharing maybe you are right and is not necessary

**Comment by alfonsobries** [comment]
@ItsANameToo you can test again ✅

File: resources/views/metadata-tags.blade.php [hunk 4]
```diff
@@ -4,11 +4,11 @@
...
+<meta name="description" content="{!! trim(htmlentities($description)) !!}" />
 
-<meta property="og:title" content="{{ trim($title) }}" />
-<meta property="og:description" content="{{ trim($description) }}" />
+<meta property="og:title" content="{!! trim(htmlentities($title)) !!}" />
+<meta property="og:description" content="{!! trim(htmlentities($description)) !!}" />
 
 @if ($image)
-    <meta property="og:image" content="{{ trim($image) }}" />
+    <meta property="og:image" content="{!! trim(htmlentities($image)) !!}" />
```
Comment (ItsANameToo) [inline resources/views/metadata-tags.blade.php:13]: no need to change this, it will always be a regular URL that we set

File: resources/views/metadata-tags.blade.php [hunk 5]
```diff
@@ -4,11 +4,11 @@
     'image' => null,
 ])
 
-<meta name="description" content="{{ trim($description) }}" />
+<meta name="description" content="{!! trim(htmlentities($description)) !!}" />
 
-<meta property="og:title" content="{{ trim($title) }}" />
-<meta property="og:description" content="{{ trim($description) }}" />
+<meta property="og:title" content="{!! trim(htmlentities($title)) !!}" />
+<meta property="og:description" content="{!! trim(htmlentities($description)) !!}" />
```
Comment (ItsANameToo) [inline resources/views/metadata-tags.blade.php:10]: so basically this turns `'` into `&#039;` and then the `!!` makes it end up as `'` again in the eventual HTML so the meta property has `'`? 🤔

Comment (alfonsobries) [reply resources/views/metadata-tags.blade.php:10]: with the changes on msq the text comes unencoded (so the `'` comes at it is). Once here the `{{ }}` directive is the one that used to encode the text as `&#039;`. 

The `htmlentities` method shows the text as it is, I suppose it uses a different way to encode the text, I would like to research the laravel directive to say it for sure. 

Quick note: the `htmlentities` method is not really necessary to fix the specific encoding issue I added it to make it secure for potential future XSS issues (as far as I can tell currently there is no way to the user to modify the metadata but this is a shared library)

**Review by ItsANameToo — APPROVED** [review]
